File: Server/patient-management-service/src/main/java/com/aarogya/patient_management_service/service/MedicalHistoryService.java
package com.aarogya.patient_management_service.service;

import com.aarogya.patient_management_service.dto.request.CreateMedicalHistoryRequest;
import com.aarogya.patient_management_service.dto.response.MedicalHistoryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface MedicalHistoryService {

    Page<MedicalHistoryResponse> getPatientMedicalHistory(String patientId, Pageable pageable);

    List<MedicalHistoryResponse> getActiveMedicalHistory(String patientId);

    MedicalHistoryResponse addMedicalHistory(String patientId, CreateMedicalHistoryRequest request);

    MedicalHistoryResponse updateMedicalHistory(String patientId, String historyId, CreateMedicalHistoryRequest request);

    void deleteMedicalHistory(String patientId, String historyId);

    List<MedicalHistoryResponse> searchMedicalHistory(String patientId, String query);

    Optional<MedicalHistoryResponse> getMedicalHistoryById(String patientId, String historyId);
}

File: Client/src/Services/Patient/medicalHistoryService.ts
import type {
  ApiResponse,
  CreateMedicalHistoryRequest,
  MedicalHistoryResponse,
} from "../../types/patientDashboard";
import { api, ENDPOINTS } from "../../utils/dashboardApi";

export class MedicalHistoryService {
  private getBasePath(patientId: string) {
    return `${ENDPOINTS.PATIENT}/${patientId}/medical-history`;
  }

  async addMedicalHistory(
    patientId: string,
    request: CreateMedicalHistoryRequest
  ): Promise<MedicalHistoryResponse> {
    try {
      const response = await api.post<ApiResponse<MedicalHistoryResponse>>(
        this.getBasePath(patientId),
        request
      );
      return response.data.data;
    } catch (error) {
      console.error(
        `Error adding medical history for patient ${patientId}:`,
        error
      );
      throw error;
    }
  }

  async updateMedicalHistory(
    patientId: string,
    historyId: string,
    request: CreateMedicalHistoryRequest
  ): Promise<MedicalHistoryResponse> {
    try {
      const response = await api.put<ApiResponse<MedicalHistoryResponse>>(
        `${this.getBasePath(patientId)}/${historyId}`,
        request
      );
      return response.data.data;
    } catch (error) {
      console.error(
        `Error updating medical history ${historyId} for patient ${patientId}:`,
        error
      );
      throw error;
    }
  }

  async deleteMedicalHistory(
    patientId: string,
    historyId: string
  ): Promise<string> {
    try {
      const response = await api.delete<ApiResponse<string>>(
        `${this.getBasePath(patientId)}/${historyId}`
      );
      return response.data.data;
    } catch (error) {
      console.error(
        `Error deleting medical history ${historyId} for patient ${patientId}:`,
        error
      );
      throw error;
    }
  }
}

export const medicalHistoryService = new MedicalHistoryService();
